src/transformers/modeling_tf_utils.py

```diff
@@ -1140,6 +1148,22 @@ def framework(self) -> str:
         """
         return "tf"
 
+    def build(self, input_shape=None):
+        if self.built or call_context().in_call:
+            self.built = True
+        else:
+            self(self.dummy_inputs, training=False)
+            self.built = True
+            self.built_with_dummies = True
+
+    def build_with_dummies(self, dummies=None):
+        if self.built_with_dummies and dummies is None:
+            return
+        if dummies is None:
+            dummies = self.dummy_inputs
+        self(dummies, training=False)
+        self.built_with_dummies = True
```

Comment:
I think the API here is a bit confusing: 
* Allowing `dummies` to essentially be any input
* Rebuilding if `dummies` is None. 
* Having `input_shape` as an argument for `build` and then not using it

I would rework to have `build` match the logic of the parent class, and `build_with_dummies` just use the model's dummy values. This way `build` remains generic and `build_with_x` specifically builds with `x` e.g. something like: 

```python
    def build(self, input_shape):
        if self.built or call_context().in_call:
            self.built = True
            return

        self(input_shape, training=False)
        self.built = True
        self.built_with_dummies = True

    def build_with_dummies(self):
        if self.built_with_dummies:
            return

        self(self.dummy_inputs, training=False)
        self.built_with_dummies = True
        self.built = True
```

This would:
* Change all the calls from `model.build()` to `model.build_with_dummies()`. We can then remove all the comments next to the build calls explaining we're using dummy inputs. 
* Remove the need to call `super().build(input_shape)` when we want the old build logic.
* Removes the need to set the input_shape to `None` in all the current build methods 

also - why have the distinction between `built` and `built_with_dummies`?